Improve Device Emulator API

aaronemassey opened the pull request:
This change simplifies and reduces the amount of boilerplate
code needed to get a device emulator up and running, thus reducing excise work
to writing tests.

Due to the change in EMUL_DEFINE and other code. This is a breaking-change for emulators.

Fixes #39486

aaronemassey committed:
emul: Simplify emulator bus initialization and API

Allow emulator creators to write an init function that can be used
across multiple busses so as to reduce the boilerplate and cognitive
load in creating an emulator.

Part of this change includes allowing access to the emul struct from a
field in a {bus}_struct api (e.g. i2c_struct), which removes the need for
sporadic usages of CONTAINER_OF to access the emul struct.

Overall, this change simplifies and reduces the amount of boilerplate
code to get a device emulator up and running, thus reducing excise work
to writing tests.

TEST=twister on accel,espi, and eeprom drivers tests

Signed-off-by: Aaron Massey <aaronmassey@google.com>

yperess commented on include/zephyr/drivers/emul.h:
```diff
@@ -8,6 +8,10 @@
 #ifndef ZEPHYR_INCLUDE_DRIVERS_EMUL_H_
 #define ZEPHYR_INCLUDE_DRIVERS_EMUL_H_
 
+#include <drivers/i2c_emul.h>
+#include <drivers/espi_emul.h>
+#include <drivers/spi_emul.h>
```
Add `zephyr/` prefix

aaronemassey replied:
Done

yperess commented on include/zephyr/drivers/espi_emul.h:
```diff
@@ -14,9 +14,8 @@
-#include <zephyr/device.h>
-#include <zephyr/drivers/espi.h>
-#include <zephyr/drivers/emul.h>
+#include <device.h>
+#include <drivers/espi.h>
```
leave the `zephyr/` prefix

aaronemassey replied:
Done

yperess commented on include/zephyr/drivers/i2c_emul.h:
```diff
@@ -21,8 +21,7 @@
 #include <zephyr/types.h>
-#include <zephyr/device.h>
-#include <zephyr/drivers/emul.h>
+#include <device.h>
```
Leave `zephyr/` prefix

aaronemassey replied:
Done

yperess commented on subsys/emul/espi/emul_espi_host.c:
```diff
@@ -147,10 +146,9 @@ static int emul_host_set_vw(struct espi_emul *emul, enum espi_vwire_signal vw,
 			    uint8_t *level)
 {
-	struct espi_host_emul_data *data;
+	struct espi_host_emul_data *data = emul->parent->data;
```
Why can't this accept the `struct emul *target`?

aaronemassey replied:
Looks like I was lacking some changes in the headers as well as clients.

Done.

yperess commented on subsys/emul/espi/emul_espi_host.c:
```diff
@@ -128,10 +128,9 @@ static int emul_host_find_index(struct espi_host_emul_data *data,
 			    uint8_t level)
 {
-	struct espi_host_emul_data *data;
+	struct espi_host_emul_data *data = emul->parent->data;
```
Same here, can we pass the `struct emul *target`?

aaronemassey replied:
Looks like I was lacking some changes in the headers as well as clients.

Done.

aaronemassey commented on subsys/emul/i2c/emul_atmel_at24.c:
```diff
@@ -133,30 +133,46 @@ static struct i2c_emul_api at24_emul_api = {
 	memset(cfg->buf, 0xff, cfg->size);
 
-	int rc = i2c_emul_register(parent, emul->dev_label, &data->emul);
+	int rc = i2c_emul_register(parent, emulator->dev_label, &data->emul);
```
Done.

aaronemassey commented on include/zephyr/drivers/spi_emul.h:
```diff
@@ -14,8 +14,8 @@
-#include <zephyr/device.h>
-#include <zephyr/drivers/emul.h>
+#include <device.h>
+#include <drivers/spi.h>
```
Done

aaronemassey commented on drivers/i2c/i2c_emul.c:
```diff
@@ -74,8 +74,8 @@ static int i2c_emul_get_config(const struct device *dev, uint32_t *dev_config)
 
-static int i2c_emul_transfer(const struct device *dev, struct i2c_msg *msgs,
-			     uint8_t num_msgs, uint16_t addr)
+static int i2c_emul_transfer(const struct device *dev, struct i2c_msg *msgs, uint8_t num_msgs,
```
clang-format

aaronemassey commented on drivers/i2c/i2c_emul.c:
```diff
@@ -119,8 +119,7 @@ static int i2c_emul_init(const struct device *dev)
 
-int i2c_emul_register(const struct device *dev, const char *name,
-		      struct i2c_emul *emul)
+int i2c_emul_register(const struct device *dev, const char *name, struct i2c_emul *emul)
```
clang-format

aaronemassey commented on include/zephyr/drivers/emul.h:
```diff
@@ -58,8 +58,7 @@ struct emul_list_for_bus {
  */
-typedef int (*emul_init_t)(const struct emul *emul,
-			   const struct device *parent);
+typedef int (*emul_init_t)(const struct emul *emul, const struct device *parent);
```
clang-format

aaronemassey commented on include/zephyr/drivers/i2c_emul.h:
```diff
@@ -69,8 +69,7 @@ typedef int (*i2c_emul_transfer_t)(const struct emul *target, struct i2c_emul *b
  */
-int i2c_emul_register(const struct device *dev, const char *name,
-		      struct i2c_emul *emul);
+int i2c_emul_register(const struct device *dev, const char *name, struct i2c_emul *emul);
```
clang-format

aaronemassey commented on include/zephyr/drivers/emul.h:
```diff
@@ -8,6 +8,10 @@
 #ifndef ZEPHYR_INCLUDE_DRIVERS_EMUL_H_
 #define ZEPHYR_INCLUDE_DRIVERS_EMUL_H_
 
+#include <zephyr/drivers/i2c_emul.h>
```
Instead of including:

Could do forward-declaration of:
		struct i2c_emul;
		struct espi_emul;
		struct spi_emul;

aaronemassey commented on include/zephyr/drivers/emul.h:
```diff
@@ -99,8 +141,7 @@ extern const struct emul __emul_list_end[];
  */
-int emul_init_for_bus_from_list(const struct device *dev,
-				const struct emul_list_for_bus *list);
+int emul_init_for_bus_from_list(const struct device *dev, const struct emul_list_for_bus *list);
```
clang-format

aaronemassey commented on include/zephyr/drivers/espi_emul.h:
```diff
@@ -13,10 +13,9 @@
-#include <zephyr/types.h>
 #include <zephyr/device.h>
 #include <zephyr/drivers/espi.h>
-#include <zephyr/drivers/emul.h>
```
Removed to prevent recursive inclusion.

yperess replied:
Is this really an issue?

aaronemassey commented on include/zephyr/drivers/espi_emul.h:
```diff
@@ -13,10 +13,9 @@
 #include <zephyr/device.h>
 #include <zephyr/drivers/espi.h>
-#include <zephyr/drivers/emul.h>
+#include <zephyr/types.h>
```
sort-lines

aaronemassey commented on include/zephyr/drivers/espi_emul.h:
```diff
@@ -33,45 +32,49 @@ extern "C" {
 struct espi_emul;
 
+/* Forward-Declare from emul.h */
+struct emul;
```
See above comment on preventing recursive inclusion. So we forward declare just this one thing.

aaronemassey commented on include/zephyr/drivers/espi_emul.h:
```diff
@@ -85,8 +88,7 @@ typedef uintptr_t (*emul_espi_api_get_acpi_shm)(struct espi_emul *emul);
  */
-typedef struct espi_emul *(*emul_find_emul)(const struct device *dev,
-					    unsigned int chipsel);
+typedef struct espi_emul *(*emul_find_emul)(const struct device *dev, unsigned int chipsel);
```
clang-format

aaronemassey commented on include/zephyr/drivers/espi_emul.h:
```diff
@@ -98,8 +100,7 @@ typedef struct espi_emul *(*emul_find_emul)(const struct device *dev,
  */
-typedef int (*emul_trigger_event)(const struct device *dev,
-				  struct espi_event *evt);
+typedef int (*emul_trigger_event)(const struct device *dev, struct espi_event *evt);
```
clang-format

aaronemassey commented on include/zephyr/drivers/espi_emul.h:
```diff
@@ -140,8 +141,7 @@ struct emul_espi_driver_api {
  */
-int espi_emul_register(const struct device *dev, const char *name,
-		       struct espi_emul *emul);
+int espi_emul_register(const struct device *dev, const char *name, struct espi_emul *emul);
```
clang-format

aaronemassey commented on include/zephyr/drivers/espi_emul.h:
```diff
@@ -154,8 +154,7 @@ int espi_emul_register(const struct device *dev, const char *name,
  */
-int emul_espi_host_send_vw(const struct device *espi_dev,
-			   enum espi_vwire_signal vw, uint8_t level);
+int emul_espi_host_send_vw(const struct device *espi_dev, enum espi_vwire_signal vw, uint8_t level);
```
clang-format

aaronemassey commented on subsys/emul/espi/emul_espi_host.c:
```diff
@@ -31,33 +31,33 @@ struct vw_data {
-	{ ESPI_VWIRE_SIGNAL_OOB_RST_ACK,   0, ESPI_SLAVE_TO_MASTER },
-	{ ESPI_VWIRE_SIGNAL_WAKE,          0, ESPI_SLAVE_TO_MASTER },
-	{ ESPI_VWIRE_SIGNAL_PME,           0, ESPI_SLAVE_TO_MASTER },
+	{ ESPI_VWIRE_SIGNAL_SLP_S3, 0, ESPI_MASTER_TO_SLAVE },
```
clang-format

carlescufi commented:
@sjg20 would you mind taking a look?

keith-zephyr commented on drivers/espi/espi_emul.c:
```diff
@@ -113,7 +113,7 @@ static int espi_emul_read_lpc_request(const struct device *dev,
 	case EACPI_GET_SHARED_MEMORY:
 		__ASSERT_NO_MSG(api->get_acpi_shm);
-		*data = (uint32_t)api->get_acpi_shm(emul);
+		*data = (uint32_t)api->get_acpi_shm(emul->parent, emul);
```
I find the use of `parent` here confusing, when the function prototype uses `target` for this parameter.

aaronemassey commented on include/zephyr/drivers/espi_emul.h:
```diff
@@ -29,6 +24,11 @@
 extern "C" {
 #endif
 
+#include <zephyr/device.h>
```
Moved inside the extern "C" clause to protect against https://stackoverflow.com/q/8526819

yperess replied:
I think this means that one of these files (or a file included from one of these) isn't using `extern "C"`

keith-zephyr commented on include/zephyr/drivers/emul.h:
```diff
@@ -70,6 +95,17 @@ extern const struct emul __emul_list_end[];
+#define Z_EMUL_REG_BUS_IDENTIFIER(_dev_node_id) (_CONCAT(_CONCAT(__emulreg_, _dev_node_id), _bus))
+
+/* Conditionally places text based on what bus _dev_node_id is on. */
+#define Z_EMUL_BUS(_dev_node_id, _i2c, _espi, _spi)                                                \
+	COND_CODE_1(DT_ON_BUS(_dev_node_id, i2c), (_i2c),                                          \
+		    (COND_CODE_1(DT_ON_BUS(_dev_node_id, espi), (_espi),                           \
+				 (COND_CODE_1(DT_ON_BUS(_dev_node_id, spi), (_spi), (-EINVAL))))))
```
Can this simplified as:
```
#define Z_EMUL_BUS(_dev_node_id)                                                \
	COND_CODE_1(DT_ON_BUS(_dev_node_id, i2c), (i2c),                                          \
		    (COND_CODE_1(DT_ON_BUS(_dev_node_id, espi), (espi),                           \
				 (COND_CODE_1(DT_ON_BUS(_dev_node_id, spi), (spi), (-EINVAL))))))
```

aaronemassey replied:
Not just used as:
`Z_EMUL_BUS(node_id, i2c, espi, spi)`

keith-zephyr commented on subsys/emul/emul_bmi160.c:
```diff
@@ -312,82 +308,28 @@ static struct i2c_emul_api bmi160_emul_api_i2c = {
-#define BMI160_EMUL_DEFINE(n, type) \
-	EMUL_DEFINE(emul_bosch_bmi160_init_##type, DT_DRV_INST(n), \
-		&bmi160_emul_cfg_##n, &bmi160_emul_data_##n)
+#define BMI160_EMUL_DEFINE(n, type, api) \
```
The `type` parameter is no longer used.

keith-zephyr commented on subsys/emul/i2c/emul_atmel_at24.c:
```diff
@@ -167,6 +180,13 @@ static int emul_atmel_at24_init(const struct emul *emul,
+/* Although we check for SPI to maintain the standard pattern of bus checking device emulators we've
+ * only integrated the I2C bus with this emulator.
+ */
+#define AT24_EMUL(n)                                                                               \
```
This macro isn't used.

aaronemassey replied:
Fixed.

keith-zephyr commented on subsys/emul/i2c/emul_atmel_at24.c:
```diff
@@ -133,30 +132,44 @@ static struct i2c_emul_api at24_emul_api = {
+	return 0;
 }
 
+#define EEPROM_AT24_EMUL_DATA(n)                                                                   \
+	struct i2c_emul emul_##n;                                                                  \
+	const struct device *i2c_##n;                                                              \
+	uint32_t cur_reg_##n;
+
+#define EEPROM_AT24_EMUL_DEFINE(n, api)                                                            \
+	EMUL_DEFINE(emul_eeprom_at24_init, DT_DRV_INST(n), &eeprom_at24_emul_cfg_##n,              \
+		    &eeprom_at24_emul_data_##n, &api)
+
+#define EEPROM_AT24_EMUL_I2C(n)                                                                    \
+	EEPROM_AT24_EMUL_DATA(n)                                                                   \
+	static const struct at_24_emul_cfg at_24_emul_cfg_##n = { .bus_label =                     \
+									  DT_INST_BUS_LABEL(n),    \
+								  .reg = at_24_emul_reg_##n,       \
+								  .addr = DT_INST_REG_ADDR(n) };   \
+	EEPROM_AT24_EMUL_DEFINE(n, &api_i2c)
```
These macros aren't used.  And line 171 has an error - it points to an incorrect reference (should be `&bus_api`)

aaronemassey replied:
Fixed.

aaronemassey commented:
Converting to draft-mode to work on comments.

aaronemassey commented:
Made a large design change here:

After a bit more digging we only use a `struct {bus}_emul` to get at `struct emul` in emulator client-code so changing all of the function signatures to work directly with `struct emul` instead. If ever needed the {bus_emul} can still be accessed via `bus->emul`

yperess commented on include/zephyr/drivers/i2c_emul.h:
```diff
@@ -20,23 +20,23 @@
 extern "C" {
 #endif
 
+#include <zephyr/device.h>
+#include <zephyr/drivers/emul.h>
+#include <zephyr/types.h>
```
Same here, this works but can't be the right solution

aaronemassey replied:
Done

yperess commented on include/zephyr/drivers/spi_emul.h:
```diff
@@ -28,15 +24,20 @@
 extern "C" {
 #endif
 
+#include <zephyr/device.h>
+#include <zephyr/drivers/emul.h>
+#include <zephyr/drivers/spi.h>
+#include <zephyr/types.h>
```
Here too

aaronemassey replied:
Done

yperess commented on subsys/emul/emul.c:
```diff
@@ -46,12 +45,49 @@ int emul_init_for_bus_from_list(const struct device *dev,
+
+		switch (emul->bus_type) {
+		case EMUL_BUS_TYPE_I2C:
+#ifdef CONFIG_I2C_EMUL
```
This `#ifdef` should be above the case, or alternatively you should move the `#endif` up so that the `break;` is still there otherwise you'll fall through

aaronemassey replied:
Oops! Good-catch! Good catch.

aaronemassey commented on include/zephyr/drivers/emul.h:
```diff
@@ -18,9 +18,23 @@
 #endif /* __cplusplus */
 
-struct device;
 struct emul;
```
@yperess 

The includes must come after the forward-declaration otherwise we get their usages before the type's been declared when we include the other headers, leading to a *real* compilation failure.

sjg20 commented:
I will take a look at this by Friday

sjg20 commented on drivers/espi/espi_emul.c:
```diff
@@ -95,6 +95,8 @@ static int espi_emul_read_lpc_request(const struct device *dev,
 	struct espi_emul *emul;
 	struct espi_emul_data *emul_data = dev->data;
 
+	ARG_UNUSED(data);
```
Move these to commit

sjg20 commented on subsys/emul/espi/emul_espi_host.c:
```diff
@@ -31,33 +31,33 @@ struct vw_data {
-	{ ESPI_VWIRE_SIGNAL_PME,           0, ESPI_SLAVE_TO_MASTER },
-	{ ESPI_VWIRE_SIGNAL_SLV_BOOT_DONE, 0, ESPI_SLAVE_TO_MASTER },
-	{ ESPI_VWIRE_SIGNAL_ERR_FATAL,     0, ESPI_SLAVE_TO_MASTER },
+	{ ESPI_VWIRE_SIGNAL_DNX_ACK, 0, ESPI_SLAVE_TO_MASTER },
```
separate commit

sjg20 commented on subsys/emul/espi/emul_espi_host.c:
```diff
@@ -233,7 +228,7 @@ uintptr_t emul_espi_host_get_acpi_shm(const struct device *espi_dev)
 	__ASSERT_NO_MSG(rc == 0);
 
-	return (uintptr_t) shm;
+	return (uintptr_t)shm;
```
whitespace change

sjg20 commented:
See also here:

https://www.kernel.org/doc/html/v4.10/process/submitting-patches.html#separate-your-changes

aaronemassey committed:
emul: Run clang-format

Run clang-format on all files touched by improved emulator API pull-request
that allowed access to the target device emulator from its bus api without
CONTAINER_OF usage.

drivers/i2c/i2c_emul.c
drivers/spi/spi_emul.c
include/zephyr/drivers/emul.h
include/zephyr/drivers/espi_emul.h
include/zephyr/drivers/i2c_emul.h
include/zephyr/drivers/spi_emul.h
subsys/emul/emul.c
subsys/emul/emul_bmi160.c
subsys/emul/espi/emul_espi_host.c
subsys/emul/i2c/emul_atmel_at24.c

TEST=twister on accel,espi, and eeprom drivers tests

Signed-off-by: Aaron Massey <aaronmassey@google.com>

aaronemassey committed:
emul: Remove or declare unused args

In several locations of the emulator code there are unused function
arguments that were never caught.

Declare these as unused or remove the unused function parameters entirely.

Signed-off-by: Aaron Massey <aaronmassey@google.com>

sjg20 commented on drivers/espi/espi_emul.c:
```diff
@@ -109,7 +109,7 @@ static int espi_emul_read_lpc_request(const struct device *dev, enum lpc_periphe
 	case EACPI_GET_SHARED_MEMORY:
 		__ASSERT_NO_MSG(api->get_acpi_shm);
-		*data = (uint32_t)api->get_acpi_shm(emul);
+		*data = (uint32_t)api->get_acpi_shm(emul->target);
```
Fine, but here is a comment for the future.

This won't work on a 64-bit machine, right? Then account neither will the function prototype as it uses 32-bit. I suppose it is not allowed to just use 'unsigned long'?

sjg20 commented on include/zephyr/drivers/espi_emul.h:
```diff
@@ -37,39 +37,39 @@ struct espi_emul;
  * @retval -EIO General input / output error.
  */
-typedef int (*emul_espi_api_set_vw)(struct espi_emul *emul, enum espi_vwire_signal vw,
+typedef int (*emul_espi_api_set_vw)(const struct emul *emulator, enum espi_vwire_signal vw,
```
Use emu for the argument name to avoid making this so long

aaronemassey replied:
Resolved in https://github.com/zephyrproject-rtos/zephyr/pull/46818

Had to create a new pull-request because GitHub won't let me update this pull-request after recreating my fork. Sorry.

aaronemassey commented:
Ugh, sorry folks. Made a terrible mistake thinking I could recreate my Zephyr fork thinking as long as the branch names matched up I could reupdate the same pull-request. Unfortunately, this is not the case.

Please move all comments and review to https://github.com/zephyrproject-rtos/zephyr/pull/46818
